starport/services/network/network.go

```diff
@@ -129,79 +130,131 @@ func (b *Builder) Blockchain(ctx context.Context, source SourceOption, options .
 	}
 	source(&o)
 
-	b.ev.Send(events.New(events.StatusOngoing, "Fetching the source code"))
+	var (
+		chainID     string
+		genesisURL  string
+		genesisHash string
+		home        = o.homePath
+		url         = o.url
+		ref         = o.ref
+		hash        = o.hash
+	)
+
+	// if a launch id is provided, chain information are fetched from Starport Network
+	if o.launchID != nil {
+		b.ev.Send(events.New(events.StatusOngoing, "Fetching chain information"))
+		chainLaunch, err := b.fetchChainLaunch(ctx, *o.launchID)
+		if err != nil {
+			return nil, err
+		}
+		b.ev.Send(events.New(events.StatusOngoing, "Chain information fetched"))
+		url = chainLaunch.SourceURL
+		hash = chainLaunch.SourceHash
+		chainID = chainLaunch.GenesisChainID
 
-	path, url, hash, err := b.fetch(ctx, o)
+		// Check if custom genesis URL is provided
+		if customGenesisURL := chainLaunch.InitialGenesis.GetGenesisURL(); customGenesisURL != nil {
+			genesisURL = customGenesisURL.Url
+			genesisHash = customGenesisURL.Hash
+		}
+
+		// If no custom home is provided, a default home determined from the launch ID is used
+		if home == "" {
+			home, err = ChainHome(*o.launchID)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	b.ev.Send(events.New(events.StatusOngoing, "Fetching the source code"))
+	path, hash, err := b.fetchSource(ctx, url, ref, hash)
 	if err != nil {
 		return nil, err
 	}
-
 	b.ev.Send(events.New(events.StatusDone, "Source code fetched"))
 
 	bc := &Blockchain{
-		appPath: path,
-		url:     url,
-		hash:    hash,
-		builder: b,
+		appPath:     path,
+		url:         url,
+		hash:        hash,
+		builder:     b,
+		genesisURL:  genesisURL,
+		genesisHash: genesisHash,
 	}
-	return bc, bc.setup(o.chainID, o.homePath, o.keyringBackend)
+	return bc, bc.setup(chainID, home, o.keyringBackend)
 }
 
-func (b *Builder) fetch(ctx context.Context, o initOptions) (path, url, hash string, err error) {
-	// determine final source configuration.
-	url = o.url
-	ref := o.ref
-	ohash := o.hash
+// fetchChainLaunch fetches the chain launch from Starport Network from a launch id
+func (b *Builder) fetchChainLaunch(ctx context.Context, launchID uint64) (launchtypes.Chain, error) {
+	res, err := launchtypes.NewQueryClient(b.cosmos.Context).Chain(ctx, &launchtypes.QueryGetChainRequest{
+		LaunchID: launchID,
+	})
+	if err != nil {
```

Comment:
```suggestion
 ```

We can just `return res.Chain, err`.

Reply:
`res` is a pointer and therefore may be `nil`